metric-anomaly-detector/src/main/java/org/hypertrace/alert/engine/metric/anomaly/detector/evaluator/BaselineRuleEvaluator.java

```diff
@@ -0,0 +1,194 @@
+            .getViolationConditionList()
+            .get(0)
+            .getDynamicThresholdCondition();
+
+    Duration duration = java.time.Duration.parse(dynamicThresholdCondition.getBaselineDuration());
+    long durationMillis = duration.toMillis();
+
+    // this query will fetch metric data, which will used for baseline calculation and evaluation
+    QueryRequest queryRequest =
+        queryRequestHandler.getQueryRequest(
+            metricAnomalyEventCondition,
+            alertTask.getTenantId(),
+            alertTask.getCurrentExecutionTime() - durationMillis,
+            alertTask.getCurrentExecutionTime());
+
+    LOGGER.debug("Rule id {}, Query request {}", alertTask.getEventConditionId(), queryRequest);
+
+    Iterator<ResultSetChunk> iterator =
+        queryRequestHandler.executeQuery(
+            Map.of(TENANT_ID_KEY, alertTask.getTenantId()), queryRequest);
+
+    List<Double> metricValuesForBaseline = new ArrayList<>();
+    List<Double> metricValuesForEvaluation = new ArrayList<>();
+
+    iterator.forEachRemaining(
+        chunk -> {
+          for (Row row : chunk.getRowList()) {
+            if (row.getColumnCount() >= 2
+                && row.getColumn(1).getValueType()
+                    == org.hypertrace.core.query.service.api.ValueType.STRING) {
+              double value = Double.parseDouble(row.getColumn(1).getString());
+              metricValuesForBaseline.add(value);
+              if (Long.parseLong(row.getColumn(0).getString())
+                  >= alertTask.getLastExecutionTime()) {
+                metricValuesForEvaluation.add(value);
+              }
+            }
+          }
+        });
+
+    LOGGER.debug(
+        "Rule id {}, Metric data for baseline {}, evaluation {}",
+        alertTask.getEventConditionId(),
+        metricValuesForBaseline,
+        metricValuesForEvaluation);
+
+    Baseline baseline =
+        getBaseline(metricValuesForBaseline.stream().mapToDouble(Double::doubleValue).toArray());
+
+    LOGGER.debug("Rule id {}, Baseline value {}", alertTask.getEventConditionId(), baseline);
+
+    int dataCount = 0, violationCount = 0;
+    for (Double metricValue : metricValuesForEvaluation) {
+      dataCount++;
+      if (metricValue < baseline.getLowerBound().getDouble()
+          || metricValue > baseline.getUpperBound().getDouble()) {
+        violationCount++;
+      }
+    }
+
+    LOGGER.debug(
+        "Rule id {}, DataCount {}, ViolationCount {}",
+        alertTask.getEventConditionId(),
+        dataCount,
+        violationCount);
+
+    if (!StaticRuleEvaluator.isViolation(dataCount, violationCount)) {
+      return Optional.empty();
+    }
+
+    return getNotificationEvent(
+        alertTask,
+        dataCount,
+        violationCount,
+        metricValuesForEvaluation,
+        baseline.getLowerBound().getDouble(),
+        baseline.getUpperBound().getDouble());
+  }
+
+  private static Baseline getBaseline(double[] metricValueArray) {
```

Comment:
Can we have a backlog ticket for re-factoring this?
The main issue (as discussed) is that now, we will have two places where the baseline is calculated.
- one in gateway service
- one is here

So, there will be chances that reported alert + inspected value in UI can differ. So, we have to make this code common either via service / via the library. The library will still have the same issue if both components running in different versions. So, service would be better (we will have to document the scaling requirement for gateway/baseline service while running with alert-engine. So, let's have a backlog ticket for both.).

Reply:
https://github.com/hypertrace/hypertrace-alert-engine/issues/54 Please add details on this